r/tests/testthat/test-Array.R

```diff
@@ -985,3 +985,14 @@ test_that("Array to C-interface", {
   delete_arrow_schema(schema_ptr)
   delete_arrow_array(array_ptr)
 })
+
+test_that("Array coverts timestamps with missing timezone /assumed local tz correctly", {
+  withr::with_envvar(c(TZ = "America/Chicago"), {
+    a <- as.POSIXct("1970-01-01 00:00:00")
+    attr(a, "tzone") <- Sys.getenv("TZ")
```

Comment:
This is a good first step, but would it be better to have two timestamps here? One that was created with `TZ` unset, and then one where we specifically set the timezone with `attr(b, "tzone"), Sys.timezone())` And confirm that those two arrays are equal?

Reply:
I've added that. But wouldn't those arrays be equal in their absolute value (without the `"tzone"` medatadata), but equally wrong (because they are both shifting the timestamp by the same amount) at the same time?

Reply:
In my mind there is some inconsistency in the Jira ticket:

> Set the timezone to local time without changing the integer value fo the timestamp. We store whatever integer R passes to us (21600), with CST as the timezone set. Display is then "1970-01-01 00:00:00 CST"

If we store the integer that R passes to us (i.e. the _absolute value_ = the number of seconds from the epoch) then the display will never be `"1970-01-01 00:00:00 CST"` in arrow, but rather the UTC date-time corresponding to the _absolute value_ with the desired timezone label, which I think it would be terribly confusing.

Reply:
> but rather the UTC date-time corresponding to the _absolute value_ with the desired timezone label, which I think it would be terribly confusing.

You need to offset for the confusing display in Arrow. Assuming that is fixed, would the result still be confusing?

Reply:
No, I think if we were to offset the display in Arrow (which I think is the solution), then the timestamp would no longer be confusing.

Reply:
The phrase "the display" here is confusing / wrong in some circumstances. When printing arrays, currently AFAICT arrow prints the timestamp in UTC for datetimes regardless if there is a timezone attached or not:

``` r
library(arrow, warn.conflicts = FALSE)

# specifically setting the timezone, and the Arrow Array repl shows UTC
ts <- as.POSIXct("2020-01-01 02:00:00", tz = "America/Chicago") + 1:10*3600
ts
#>  [1] "2020-01-01 03:00:00 CST" "2020-01-01 04:00:00 CST"
#>  [3] "2020-01-01 05:00:00 CST" "2020-01-01 06:00:00 CST"
#>  [5] "2020-01-01 07:00:00 CST" "2020-01-01 08:00:00 CST"
#>  [7] "2020-01-01 09:00:00 CST" "2020-01-01 10:00:00 CST"
#>  [9] "2020-01-01 11:00:00 CST" "2020-01-01 12:00:00 CST"
attr(ts, "tzone")
#> [1] "America/Chicago"

arr <- Array$create(ts)
arr
#> Array
#> <timestamp[us, tz=America/Chicago]>
#> [
#>   2020-01-01 09:00:00.000000,
#>   2020-01-01 10:00:00.000000,
#>   2020-01-01 11:00:00.000000,
#>   2020-01-01 12:00:00.000000,
#>   2020-01-01 13:00:00.000000,
#>   2020-01-01 14:00:00.000000,
#>   2020-01-01 15:00:00.000000,
#>   2020-01-01 16:00:00.000000,
#>   2020-01-01 17:00:00.000000,
#>   2020-01-01 18:00:00.000000
#> ]
arr$type$timezone()
#> [1] "America/Chicago"

as.vector(arr)
#>  [1] "2020-01-01 03:00:00 CST" "2020-01-01 04:00:00 CST"
#>  [3] "2020-01-01 05:00:00 CST" "2020-01-01 06:00:00 CST"
#>  [5] "2020-01-01 07:00:00 CST" "2020-01-01 08:00:00 CST"
#>  [7] "2020-01-01 09:00:00 CST" "2020-01-01 10:00:00 CST"
#>  [9] "2020-01-01 11:00:00 CST" "2020-01-01 12:00:00 CST"
attr(as.vector(arr), "tzone")
#> [1] "America/Chicago"


# without setting the timezone, and the Arrow Array repl still shows UTC
ts <- as.POSIXct("2020-01-01 02:00:00") + 1:10*3600
ts
#>  [1] "2020-01-01 03:00:00 CST" "2020-01-01 04:00:00 CST"
#>  [3] "2020-01-01 05:00:00 CST" "2020-01-01 06:00:00 CST"
#>  [5] "2020-01-01 07:00:00 CST" "2020-01-01 08:00:00 CST"
#>  [7] "2020-01-01 09:00:00 CST" "2020-01-01 10:00:00 CST"
#>  [9] "2020-01-01 11:00:00 CST" "2020-01-01 12:00:00 CST"
attr(ts[[1]], "tzone")
#> NULL

arr <- Array$create(ts)
arr
#> Array
#> <timestamp[us]>
#> [
#>   2020-01-01 09:00:00.000000,
#>   2020-01-01 10:00:00.000000,
#>   2020-01-01 11:00:00.000000,
#>   2020-01-01 12:00:00.000000,
#>   2020-01-01 13:00:00.000000,
#>   2020-01-01 14:00:00.000000,
#>   2020-01-01 15:00:00.000000,
#>   2020-01-01 16:00:00.000000,
#>   2020-01-01 17:00:00.000000,
#>   2020-01-01 18:00:00.000000
#> ]
arr$type$timezone()
#> [1] ""

as.vector(arr)
#>  [1] "2020-01-01 03:00:00 CST" "2020-01-01 04:00:00 CST"
#>  [3] "2020-01-01 05:00:00 CST" "2020-01-01 06:00:00 CST"
#>  [5] "2020-01-01 07:00:00 CST" "2020-01-01 08:00:00 CST"
#>  [7] "2020-01-01 09:00:00 CST" "2020-01-01 10:00:00 CST"
#>  [9] "2020-01-01 11:00:00 CST" "2020-01-01 12:00:00 CST"
attr(as.vector(arr), "tzone")
#> NULL
```

But as I showed up there, when pulling the data back in with `as.vector(arr)`, the timezone is pulled in with it so that when R displays the timestamp it is faithful to the original timestamp.

Reply:
> The phrase "the display" here is confusing / wrong in some circumstances. When printing arrays, currently AFAICT arrow prints the timestamp in UTC for datetimes regardless if there is a timezone attached or not

Yes, that's what I meant. And indeed this is only for the _arrow_ display / pretty printing.

I think the relevant issue for this is https://issues.apache.org/jira/browse/ARROW-14567

Reply:
Thanks for digging that up, I _assumed_ it existed already but hadn't gone searching

Reply:
👍🏻 I think we are all on the same page. In this situation what is the solution? I think we can attach the local / system timezone when it isn't passed explicitly (and this would theoretically solve this Jira issue). Should we then open a follow-up issue regarding the printing of arrow timestamps in R (which falls outside the scope of this ticket)? Which would be an alignment with [ARROW-14567](https://issues.apache.org/jira/browse/ARROW-14567)?

Reply:
For now you could use `strftime` in R to print localized times.